ttnctl/cmd/applications_onjoinregistration.go

```diff
@@ -0,0 +1,194 @@
+
+		conn, manager := util.GetHandlerManager(ctx, appID)
+		defer conn.Close()
+
+		var onJoinSetting bool
+		settingStr := args[0]
+		if settingStr == "on" {
+			onJoinSetting = true
+		} else if settingStr == "off" {
+			onJoinSetting = false
+		} else {
+			// Incorrect string
+			ctx.Fatal("Please choose \"on\" or \"off\" for the On-Join Registration setting.")
+		}
+
+		resultCtx := ctx.WithField("Setting", settingStr)
+
+		app, err := manager.GetApplication(appID)
+		if err != nil {
+			resultCtx.WithError(err).Fatal("Couldn't retrieve application information")
+		}
+		if app.OnJoinRegistration == onJoinSetting {
+			resultCtx.Fatal("On-join registration setting is already set")
+		}
+
+		if onJoinSetting {
+			ctx.Info("Generating handler-stored access key")
+			accessKeyName = fmt.Sprintf("%s-%s-%s", onJoinRegistrationAccessKeyName, appID, randStr(30))
+			accessKey, err = account.AddAccessKey(appID, accessKeyName, []types.Right{rights.Devices})
+			if err != nil {
+				ctx.WithError(err).Fatal("Couldn't create access key on the account server")
+			}
+			defer func() {
+				if !success {
+					account.RemoveAccessKey(appID, accessKey.Name)
+				}
+			}()
+
+			if len(args) >= 2 {
+				// EUI filled in as argument
+				eui = new(types.AppEUI)
+				err := eui.UnmarshalText([]byte(args[1]))
+				if err != nil {
+					ctx.WithError(err).Fatal("Couldn't read given EUI")
+				}
+
+				app, err := account.FindApplication(appID)
+				if err != nil {
+					ctx.WithError(err).Fatal("Could not find application")
+				}
+
+				var euiInStorage = false
+				for _, storedEUI := range app.EUIs {
+					if *eui == storedEUI {
+						// Found EUI
+						euiInStorage = true
+						break
+					}
+				}
+				if !euiInStorage {
+					err = account.AddEUI(appID, *eui)
+					if err != nil {
+						ctx.WithError(err).Fatal("Couldn't add EUI to the account server")
+					}
+					defer func() {
+						if !success {
+							account.RemoveEUI(appID, *eui)
+						}
+					}()
+				}
+			} else {
+				// EUI not filled in
+				eui, err = account.GenerateEUI(appID)
+				if err != nil {
+					ctx.WithError(err).Fatal("Couldn't generate EUI")
+				}
+				defer func() {
+					if !success {
+						account.RemoveEUI(appID, *eui)
+					}
+				}()
+			}
+			bytesEUI, err := eui.MarshalText()
+			if err == nil {
+				resultCtx = resultCtx.WithField("AppEUI", string(bytesEUI[:]))
+			}
+
+			if len(args) >= 3 {
+				// AppKey filled in
+				appKey = new(types.AppKey)
+				err = appKey.UnmarshalText([]byte(args[2]))
+				if err != nil {
+					ctx.WithError(err).Fatal("Couldn't read given AppKey")
+				}
+			} else {
+				// AppKey not filled in
+				appKey = new(types.AppKey)
+				random.FillBytes(appKey[:])
+			}
+			bytesKey, err := appKey.MarshalText()
+			if err == nil {
+				resultCtx = resultCtx.WithField("AppKey", string(bytesKey[:]))
+			}
+		} else {
+			accessKeyName = app.OnJoinRegistrationAccessKeyName
+			defer func() {
+				if success {
```

Comment:
If on-join registration deactivation was successful on the handler side, `ttnctl` removes the access key. If it wasn't successful, we keep the access key active, so that we don't end up with an on-join registration that is considered enabled, but is actually broken.